pss/pss_test.go

```diff
@@ -495,6 +495,133 @@ func TestAddressMatchProx(t *testing.T) {
 	}
 }
 
+func TestMessageEnqueued(t *testing.T) {
+
+	// setup
+	privkey, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	addr := make([]byte, 32)
+	addr[0] = 0x01
+	ps := newTestPssStart(privkey, network.NewKademlia(addr, network.NewKadParams()), NewParams(), false)
+	outboxCapacity := 2
+	ps.outbox = make(chan *outboxMsg, outboxCapacity)
+
+	dequeue := func() *PssMsg {
+		select {
+		case oMsg := <-ps.outbox:
+			return oMsg.msg
+		default:
+			return nil
+		}
+	}
+
+	messageAddr := addr
+	topic := [4]byte{}
+	data := []byte{0x66, 0x6f, 0x6f}
+
+	msg := testMessage(messageAddr, topic, data)
+
+	err = ps.enqueue(msg, false)
+	if err != nil && ps.forwardPending != 1 {
+		t.Fatal("Message should have space for enqueuing")
+	}
+	t.Log("Message enqueued", "Outbox len", len(ps.outbox))
+
+	oMsg := dequeue()
+	if oMsg == nil {
+		t.Fatal("Message should have been dequeued")
+	}
+	if ps.forwardPending != 1 {
+		t.Fatal("Pending message should still be 1")
+	}
+}
+
+// Test a race condition enqueueing two new messages and a pending message in an outbox of capacity 2
+func TestMessageEnqueuedRaceCondition(t *testing.T) {
+
+	// setup
+	privkey, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	addr := make([]byte, 32)
+	addr[0] = 0x01
+	ps := newTestPssStart(privkey, network.NewKademlia(addr, network.NewKadParams()), NewParams(), false)
+	outboxCapacity := 2
+	ps.outbox = make(chan *outboxMsg, outboxCapacity)
+
+	topic := [4]byte{}
+	data := []byte{0x66, 0x6f, 0x6f}
+
+	finish := make(chan struct{})
+	var errorCount int
+	var enqueuedPending bool
+	newMsg := func(iteration int) {
+		err := ps.enqueue(testMessage(addr, topic, data), false)
+		if err != nil {
+			log.Debug("Expected error enqueued for new message", "iteration", iteration)
+			errorCount++
+		} else {
+			log.Debug("Enqueued new message", "Outbox length", len(ps.outbox), "iteration", iteration)
+		}
+		finish <- struct{}{}
+	}
+
+	reEnqueue := func(iteration int) {
+		time.Sleep(1 * time.Millisecond)
+		err := ps.enqueue(testMessage(addr, topic, data), true)
+		if err != nil {
+			enqueuedPending = false
+			log.Error("Unexpected error for pending message", "iteration", iteration)
+		} else {
+			enqueuedPending = true
+			log.Debug("Enqueued pending message", "Outbox length", len(ps.outbox), "iteration", iteration)
+		}
+		finish <- struct{}{}
+	}
+
+	for iteration := 0; iteration < 300; iteration++ {
+		log.Debug("------New Iteration-----", "iteration", iteration)
+		ps.forwardPending = 1
+		ps.outbox = make(chan *outboxMsg, outboxCapacity)
+		enqueuedPending = false
+		errorCount = 0
+
+		go newMsg(iteration)
+		go newMsg(iteration)
+		go reEnqueue(iteration)
```

Comment:
This seems to me like a bit of a roundabout way of achieving the test.

May I suggest making the `forward` method in `pss/pss.go` replacable, then have a component in the test that lets you hold and re-queue messages one at a time? This way you can do this test while the actual pss service is running `ps.Start()`

Reply:
The point of this test is check that we can't have a race condition when two new messages are received at the same time thanks to the `outboxMutex`. In order to test it (or provoke it removing the mutes) is adding two messages at the same time and that they both check the condition on line 569 (`pendingSize < cap(p.outbox`) before the other increase the pending messages. It is not related with the "forward" execution. That's why I repeat that test n times (if you remove the mutex you can see it does not always fail in the first iteration).
I know is a tricky situation, but I didn't find a better way to test that race condition.

I can remove the third go routine, as the reEnqueue could be executed afterwards.

Reply:
We don't really have to test the mutex? We rely on that it works, no? :)

if we have a pluggable forward function instead you can control the amount of pending messages you have at any one time. From the perspective of new messages being added, the current capacity of the queue is the same whether existing messages are attempted being sent (pending) or whether they are still just occupying a space in the queue. Do we need to measure anything else than the effect, really?

Furthermore, if the mutex makes it hard to write a sensible test, then that's another argument in favor of @zelig 's approach IMO.

Reply:
Well, I didn't want to test the mutex, I wanted to test (before adding the mutex) that it solved the problem. Maybe it is too obvious and we don't need that test at all.

I agree that mutex makes hard to test, and I think @zelig's solution is better overall. The only point in favour of this solution is that it changed less code and it was (almost) ready.

BTW, the forward function does not control the pending messages. We changed the meaning of `forwardPending` variable from first commit. I wrote it in the pull request description. Now `forwardPending` means the number of messages in the outbox or being forwarded and is only decreased when a message successfully or not leaves the system.

Reply:
Yes and I realize what I was writing doesn't apply because this is not about popping the queue but inserting, so never mind %)